adr/0011-python-application-deployment.md

```diff
@@ -0,0 +1,75 @@
+authors. Some of these authors have moved on to different organisations and some only worked at eLife temporarily.
+
+Each author has their own preferred tools used for development. The deployment of this code on to production servers
+is the result of using the author's preferred tools in a CI/CD pipeline to create the project's environment, install the 
+project's dependencies, pre-test tasks like linting or complexity checking, testing and post-testing tasks like coverage
+checks. These tasks are often glued together by the CI runner, or bash scripts or other python code and, recently, Make 
+files.
+
+In the case of projects with multiple authors, there are often multiple ways present to achieve the same result that 
+suits individual authors best.
+
+Python projects have traditionally left deployment and configuration up to the project formula. The project formula
+describes the requirements of the environment the project lives in, for example, does a database or web server need to 
+exist? What directories do log files live in and what permissions do they need? These and dozens of other similar cases
+are handled by the formula on behalf of the Python project. The project author doesn't necessarily need to know about 
+how these cases are solved, the formula is built and maintained by those building and maintaining the systems these
+projects run on in production.
+
+Python projects require a Python runtime and Python packages can be installed system-wide or locally using a tool called
+'virtualenv', now officially part of Python3. Dependency and runtime management has become complicated in the move from
+Python2 to Python3 and the introduction of new language features and dependency management tools; the OS distributions 
+used in production use a fixed version of the Python runtime, often with features missing or mutually exclusive with 
+other features (Ubuntu + Python3 + pip + venv for example). Attempts to accommodate the latest runtimes has introduced
+instability and complexity in the rest of the system.
+
+Containerisation has made describing and building Python applications much easier for a project's author and the line 
+between traditional code development and code deployment is less clear. Container descriptions often live within the 
+code repository instead of in a separate one. Containers serve a single purpose, running a single process, but are 
+composable so that a cluster of containers may be described: one for the application, another for a webserver, another
+for a database and so on.
+
+Deploying a container that has the project within perfectly installed and configured is very similar to deploying a fat
+application binary: a single discrete artifact with all dependencies accounted for. Deployment consists of dropping the
+artifact into the runtime and reloading a service.
+
+Because container descriptions live alongside the codebase and are used during development they suffer the same 
```

Comment:
Seems like everyone is an agreement that containerisation in general is preferable. (And as has been pointed out we were already doing)
How exactly that works in development might be the only question. Do we only support development completely through docker or do we also support a combination (e.g. use containers for other services, but allow local development for the main project - while also being able to test everything deployed as containers). For the projects I authored ([example](https://github.com/elifesciences/data-pipeline-ejp-to-json-converter)), I have so far supported both. In the end, containerisation is not the goal, but making our lives easier and solve problems (good points above). 
I'm personally not at all against using a container for the main project for development. But I do care that it doesn't actually make development more difficult for me or ties me to a particular IDE/editor. Hopefully we don't need to solve it in every single project but can find a general solution. It probably doesn't need to be solved up-front either but would be nice if potential issues are considered. (Might be in particular useful for Libero projects where we hope to get other developers on board)

Reply:
I 100% agree. I have always been of the opinion that developers know how to develop best and should be left to pick the best tools for themselves. I absolutely do not want to impose a particular philosophy of programming or mandate what IDE is be used. This ADR is concerned with a slice of development (testing and linting) and a slice of CI/CD (building and deployment). The two worlds are overlapping because that's the nature of containerisation and container descriptions living alongside project code. To a lesser extant, the convenience of a consistent process for reproducing code environments and testing projects between team members would be handy and a time saver.

I don't use Docker or an IDE for development. My philosophy around programming is old fashioned and definitely anti-social. When we need to coordinate, however, then a container providing a fixed, predictable and familiar environment will help bridge that. Same goes for the interface. @GiancarloFusiello is proposing a Make file that will handle that. I would suggest that that Make file concern itself solely with building, installing and testing things within the container. If it's good at what it does then it will naturally displace what we use for development also.